nemo/collections/common/data/lhotse/nemo_adapters.py

```diff
@@ -274,58 +276,76 @@ def _validate(self) -> None:
+                    raw_audio = tar.extractfile(tar_info).read()
+                    yield data, raw_audio, tar_info
+
     def __iter__(self) -> Generator[Cut, None, None]:
         shard_ids = self.shard_ids
 
         if self.shuffle_shards:
             seed = resolve_seed(self.shard_seed)
             random.Random(seed).shuffle(shard_ids)
 
-        for sid in shard_ids:
-            manifest_path = self.paths[sid] if len(self.paths) > 1 else self.paths[0]
-            shard_manifest = {data["audio_filepath"]: data for data in self.shard_id_to_manifest[sid]}
-            tar_path = self.shard_id_to_tar_path[sid]
-            if self.random_access:
-                tar_file = tarfile.open(fileobj=BytesIO(open_best(tar_path, mode="rb").read()), mode="r")
-            else:
-                tar_file = tarfile.open(fileobj=open_best(tar_path, mode="rb"), mode="r|*")
-            with tar_file as tar:
-                iterable_tar = iter(tar)
-                for data in shard_manifest:
-                    if self.random_access:
-                        tar_info = tar.getmember(data["audio_filepath"])
-                    else:
-                        tar_info = next(iterable_tar)
-                        assert (
-                            data["audio_filepath"] == tar_info.name
-                        ), f"Mismatched JSON manifest and tar file. {data['audio_filepath']=} != {tar_info.name=}"
-
-                    raw_audio = tar.extractfile(tar_info).read()
-                    # Note: Lhotse has a Recording.from_bytes() utility that we won't use here because
-                    #       the profiling indicated significant overhead in torchaudio ffmpeg integration
-                    #       that parses full audio instead of just reading the header for WAV files.
-                    # recording = lhotse.Recording.from_bytes(raw_audio, recording_id=tar_info.path)
-                    meta = soundfile.info(BytesIO(raw_audio))
-                    recording = Recording(
-                        id=tar_info.path,
-                        sources=[AudioSource(type="memory", channels=list(range(meta.channels)), source=raw_audio)],
-                        sampling_rate=int(meta.samplerate),
-                        num_samples=meta.frames,
-                        duration=meta.duration,
-                    )
-                    cut = recording.to_cut()
-                    cut.supervisions.append(
-                        SupervisionSegment(
-                            id=cut.id,
-                            recording_id=cut.recording_id,
-                            start=0,
-                            duration=cut.duration,
-                            text=data.get(self.text_field),
-                            language=data.get(self.lang_field),
-                        )
-                    )
-                    cut.custom = _to_custom_attr_dict(data)
-                    yield cut
+        if self.tarred_random_access:
```

Comment:
I'd suggest 

```python
iter_fn = self._iter_random_read if self.tarred_random_access else self._iter_sequential

for data, raw_audio, tar_info in iter_fn(shard_ids):
    meta = soundfile.info(BytesIO(raw_audio))
        recording = Recording(
            id=tar_info.path,
            sources=[AudioSource(type="memory", channels=list(range(meta.channels)), source=raw_audio)],
            sampling_rate=int(meta.samplerate),
            num_samples=meta.frames,
            duration=meta.duration,
        )
        cut = recording.to_cut()
        cut.supervisions.append(
            SupervisionSegment(
                id=cut.id,
                recording_id=cut.recording_id,
                start=0,
                duration=cut.duration,
                text=data.get(self.text_field),
                language=data.get(self.lang_field),
            )
        )
        cut.custom = _to_custom_attr_dict(data)
        return cut
```

(since `create_recording` is unlikely to be re-used; but if you still want to leave it I'd at least rename to `create_cut`).

Reply:
Changed to suggested code.